Node registration changes master

detiber opened the pull request:
- Add openshift_facts role and module
  - Created new role openshift_facts that contains an openshift_facts module
  - Refactor openshift_* roles to use openshift_facts instead of relying on
    defaults
  - Refactor playbooks to use openshift_facts
  - Cleanup inventory group_vars

- Add README_OSE.md and byo playbooks

- Update defaults
  - update openshift_master role firewall defaults
    - remove etcd peer port, since we will not be supporting clustered embedded
      etcd
    - remove 8444 since console now runs on the api port by default
    - add 8444 and 7001 to disabled services to ensure removal if updating

- Add new role os_env_extras_node that is a subset of the docker role
  - previously, we were starting/enabling docker which was causing issues with some
  installations
  - Does not install or start docker, since the openshift-node role will
    handle that for us
  - Only adds root to the dockerroot group
  - Update playbooks to use ops_env_extras_node role instead of docker role

- os_firewall bug fixes
  - ignore ip6tables for now, since we are not configuring any ipv6 rules
  - if installing package do a daemon-reload before starting/enabling service

- Add aws support to bin/cluster

- Add list action to bin/cluster

- Add update action to bin/cluster

- cleanup some stray debug statements

- some variable renaming for clarity

twiest commented on roles/os_env_extras_node/tasks/main.yml:
```diff
@@ -0,0 +1,7 @@
+---
+- copy: src=enter-container.sh dest=/usr/local/bin/enter-container.sh mode=0755
```
we should be using 'docker exec' and not enter-container. The reason for this is selinux. When you use nsenter, and change things inside of the running container, container processes will start getting selinux denials when accessing whatever you changed.

This script looks like one I wrote awhile ago. If this is mine, I think we should just nuke the script entirely.

detiber replied:
Consider it axed, thanks...  Do we still need the root user as part of the dockerroot group? As far as I can tell it is meant for giving non-root users access to the docker daemon. @jwhonce do you know?

jwhonce replied:
You are correct.  I kept it since there was some error generated if dockerroot was not found.

twiest commented on playbooks/gce/openshift-node/config.yml:
```diff
@@ -1,49 +1,123 @@
-  hosts: localhost
-  gather_facts: no
+- name: Gather facts for hosts to configure
+  hosts: tag_env-host-type-{{ oo_env }}-openshift-node
```
This is going to gather facts for all nodes, not just the ones to be configured.

In a large environment, this will take a long time. Do we actually need to run this on all nodes?

Could we instead change it to this?
```
  hosts: oo_nodes_to_config
```

detiber replied:
My first attempt was to use oo_nodes_to_config, and I ran into issues with variable precedence causing variables to not be templated correctly.  I'll play around with the variable definitions a bit to see if I can sort them out.

twiest commented on playbooks/gce/openshift-node/config.yml:
```diff
@@ -1,49 +1,123 @@
     when: oo_host_group_exp is defined
+  - name: Find masters for env
+    add_host: "name={{ item }} groups=oo_masters_for_node_config"
+    with_items: groups['tag_env-host-type-' + oo_env + '-openshift-master']
```
I don't see oo_masters_for_node_config used anywhere.

Is this section used?

detiber replied:
Good catch, it is not. See previous note about variable precedence and oo_nodes_to_config. If I can sort out the variable precedence issues, then I'll go back to using oo_masters_for_node_config, since it cleans up readability for later references to master hosts.

twiest commented on playbooks/gce/openshift-node/config.yml:
```diff
@@ -1,49 +1,123 @@
+      dest: "{{ mktemp.stdout }}"
+
+# TODO: sync generated certs between masters
+#
```
Shouldn't the masters do this themselves instead of relying on an external config management?

Like maybe they store these in etcd or something?

detiber replied:
Unless I'm mistaken, the system will attempt to generate the certs when coming up if they do not exist (including the ca certificate). We could probably get away with only distributing the ca certs, but I would have to defer to the upstream devs.

If we can get node registration working only for oo_nodes_to_config, then we can get away without worrying about distributing the node certs to the other master hosts as well.

@deads2k, will additional masters attempt to pull existing master certs?

deads2k replied:
> will additional masters attempt to pull existing master certs?

Today, no.  We've kicked around the idea of having the config stored in etcd, but I don't think that's going to be done in the short term unless there are very strong requirements.  For now, we'll be referencing files on disk.

As for us generating the certificates.  I think that in any production environment, openshift won't have a signer cert/key pair, so you'll be running with `--create-certs=false` since we won't be able to mint any certificates.  If you do not set `--create-certs=false` and you are not running with a config file, we will create a signer cert and create all of the additional certificates we want.

twiest commented on roles/openshift_node/tasks/main.yml:
```diff
@@ -47,42 +49,10 @@
-# Always bounce service to pick up new credentials
 - name: Start and enable openshift-node
-  service: name=openshift-node enabled=yes state=restarted
+  service: name=openshift-node enabled=yes state=started
```
love this fix!

detiber commented:
@jwhonce Still some more cleanup to do, but I was able to successfully deploy a GCE cluster using bin/cluster and `openshift ex router --create` worked.

jwhonce commented on cluster.sh:
```diff
@@ -65,6 +65,10 @@ function check_argval {
+
+    if [[ $1 == 'gce' ]]; then
+        echo "Use bin/cluster for GCE clusters now"
+    fi
```
Why not just exit here after echo?

detiber commented:
@twiest @jwhonce please review again.

jwhonce commented on playbooks/aws/openshift-node/terminate.yml:
```diff
@@ -0,0 +1 @@
+../openshift-master/terminate.yml
```
This looks odd to have node playbook just call master playbook.  If not an error, you may want to add comment.

detiber replied:
@jwhonce ah, the diff is a bit misleading, it is actually a symlink.  In reality, I could probably do something like openshift-cluster/launch.yml and openshift-cluster/launch-instances.yml, I was just trying to replicate the gce playbooks as much as possible.

jwhonce commented:
@detiber With the module python getting more complex, you should at least have a issue for creating unit tests for the code.  TIA

detiber commented:
@jwhonce I went ahead and created issues for the testing discussed.

jwhonce commented:
:+1:

detiber commented:
While testing further changes to update the aws playbooks for using ips instead of hostnames, I've discovered some oddities around how the api and console urls are being configured currently. Cleaning that up now.

twiest commented on README_AWS.md:
```diff
@@ -51,7 +51,11 @@ OSX:
 --------------
 1. cd openshift-ansible
-1. Try to list all instances:
+1. To create a cluster with one master and two nodes:
```
So the point of this section is to test that you have the creds setup right.

I'd prefer a read-only test (like what "./cloud.rb aws list" is).

I think it's a bit of overkill to have to create a cluster in order to test your setup.

detiber replied:
Agreed, I'll see about pulling https://github.com/openshift/openshift-ansible/pull/125 in as well so there is a list action we can use for testing.

twiest commented on README_GCE.md:
```diff
@@ -65,12 +65,11 @@ Install Dependencies
 --------------
 1. cd openshift-ansible/
-2. Try to list all instances:
+1. To create a cluster with one master and two nodes
```
Same as my comment in aws instructions

twiest commented on playbooks/aws/openshift-cluster/launch.yml:
```diff
@@ -0,0 +1,60 @@
+    - name: Generate master instance names(s)
+      set_fact: scratch={{ cluster_id }}-{{ k8s_type }}-{{ '%05x' |format( 1048576 |random) }}
+      register: master_names_output
+      with_sequence: start=1 end={{ masters }}
```
I assume masters is a number by how it's used. Should we call it num_masters instead?

I'm not sure where this name is being set. To be clear, not a blocker to merge.

twiest commented on playbooks/aws/openshift-cluster/launch.yml:
```diff
@@ -0,0 +1,60 @@
+    - name: Generate node instance names(s)
+      set_fact: scratch={{ cluster_id }}-{{ k8s_type }}-{{ '%05x' |format( 1048576 |random) }}
+      register: node_names_output
+      with_sequence: start=1 end={{ nodes }}
```
Same comment as 'masters' above, but for 'nodes' (num_nodes?).

twiest commented on playbooks/aws/openshift-master/terminate.yml:
```diff
@@ -0,0 +1,55 @@
+      ec2_tag: >
+        resource={{ item.item.item.ec2_id }}
+        region={{ item.item.item.ec2_region }}
+        state=present
```
We prefer not to use the yaml '>' as it causes unexpected newline issues sometimes.

Instead, consider using this (untested):
```
      ec2_tag:
        resource: "{{ item.item.item.ec2_id }}"
        region: "{{ item.item.item.ec2_region }}"
        state: present
```

detiber replied:
That was my first attempt, however you cannot use dictionary arguments with the args keyword, I can however just make it a long line and remove the '>'

twiest commented on roles/openshift_facts/library/openshift_facts.py:
```diff
@@ -0,0 +1,414 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
```
To help keep consistency, we add this line to the top of all python files:
```
# vim: expandtab:tabstop=4:shiftwidth=4
```

detiber commented:
@twiest This PR now correctly sets the internal and external hostnames to ip addresses now and I believe I've also addressed all of the issues you've raised.

@jwhonce I've pulled in the changes that you've made in https://github.com/openshift/openshift-ansible/pull/125, with some minor modifications (basically, you can get bin/cluster to do a list all by running `bin/cluster list <provider> ''`, and adding aws support). I've also added rudimentary update support as well.

If everything looks good, then I can rebase and squash the commits for merging.

jwhonce commented on playbooks/aws/openshift-cluster/launch.yml:
```diff
@@ -0,0 +1,60 @@
+
+- include: ../openshift-node/config.yml
+  vars:
+    oo_host_group_exp: "groups[\"tag_env-host-type_{{ cluster_id }}-openshift-node\"]"
```
Please add to the end of the file:

- include: list.yml

jwhonce commented:
@detiber If you add that one line to the end of the launch.yml scripts, I can close https://github.com/openshift/openshift-ansible/pull/125

Otherwise, LGTM

detiber commented:
@jwhonce done

twiest commented on README_AWS.md:
```diff
@@ -50,8 +50,28 @@ OSX:
 --------------
-1. cd openshift-ansible
-1. Try to list all instances:
+1. cd ost all instances
```
I don't understand this line. This seems to be a copy / paste bug.

twiest commented on README_AWS.md:
````diff
@@ -50,8 +50,28 @@ OSX:
+1. cd ost all instances
 ```
-  ./cloud.rb aws list
+  bin/cluster list aws ''
````
Why the '' at the end?

detiber replied:
Basically, the list playbook checks if cluster_id is an empty string, and if so it'll list all instances (which is a better test than asking the user to already have a defined cluster before running list as a test).  I'll add a comment to indicate that is what is being done.

twiest commented on README_GCE.md:
````diff
@@ -65,12 +65,28 @@ Install Dependencies
+1. Try to list all instances
 ```
-  ./cloud.rb gce list
+  bin/cluster list gce ''
````
same '' question

twiest commented:
@detiber Looking really good. I think this is ready to be rebased, squashed and merged.

If you want to address my comments above when you do the rebase and squash, that'd be great. Otherwise, a separate PR is fine too.

When you want this merged, please let either @jwhonce or I know and we'll get it in.

Thanks, this is a great PR!

detiber commented:
@twiest issues addressed, rebased and squashed. Many thanks for putting up with the huge changeset.

twiest commented:
@detiber LGTM, :+1: 


@jwhonce Since this is such a big change, I'd like your +1 of the final PR before merging. If you +1 it, please merge it too.
